statesync/reactor.go

```diff
@@ -27,189 +62,299 @@ const (
+// NewReactor returns a reference to a new state-sync reactor. It accepts a Context
+// which will be used for cancellations and deadlines when executing the reactor,
+// a p2p Channel used to receive and send messages, and a router that will be
+// used to get updates on peers.
+func NewReactor(
+	logger log.Logger,
+	conn proxy.AppConnSnapshot,
+	connQuery proxy.AppConnQuery,
+	snapshotCh, chunkCh *p2p.Channel,
+	peerUpdates p2p.PeerUpdates,
+	tempDir string,
+) *Reactor {
 	r := &Reactor{
-		conn:      conn,
-		connQuery: connQuery,
+		conn:            conn,
+		connQuery:       connQuery,
+		snapshotCh:      snapshotCh,
+		snapshotChDone:  make(chan bool),
+		chunkCh:         chunkCh,
+		chunkChDone:     make(chan bool),
+		peerUpdates:     peerUpdates,
+		peerUpdatesDone: make(chan bool),
+		tempDir:         tempDir,
 	}
-	r.BaseReactor = *p2p.NewBaseReactor("StateSync", r)
-	return r
-}
 
-// GetChannels implements p2p.Reactor.
-func (r *Reactor) GetChannels() []*p2p.ChannelDescriptor {
-	return []*p2p.ChannelDescriptor{
-		{
-			ID:                  SnapshotChannel,
-			Priority:            3,
-			SendQueueCapacity:   10,
-			RecvMessageCapacity: snapshotMsgSize,
-		},
-		{
-			ID:                  ChunkChannel,
-			Priority:            1,
-			SendQueueCapacity:   4,
-			RecvMessageCapacity: chunkMsgSize,
-		},
-	}
+	r.BaseService = *service.NewBaseService(logger, "StateSync", r)
+	return r
 }
 
-// OnStart implements p2p.Reactor.
+// OnStart starts separate go routines for each p2p Channel and listens for
+// envelopes on each. In addition, it also listens for peer updates and handles
+// messages on that p2p channel accordingly. The caller must be sure to execute
+// OnStop to ensure the outbound p2p Channels are closed. No error is returned.
 func (r *Reactor) OnStart() error {
+	// Listen for envelopes on the snapshot p2p Channel in a separate go-routine
+	// as to not block or cause IO contention with the chunk p2p Channel. Note,
+	// we do not launch a go-routine to handle individual envelopes as to not
+	// have to deal with bounding workers or pools.
+	go r.processSnapshotCh()
+
+	// Listen for envelopes on the chunk p2p Channel in a separate go-routine
+	// as to not block or cause IO contention with the snapshot p2p Channel. Note,
+	// we do not launch a go-routine to handle individual envelopes as to not
+	// have to deal with bounding workers or pools.
+	go r.processChunkCh()
+
+	go r.processPeerUpdates()
+
 	return nil
 }
 
-// AddPeer implements p2p.Reactor.
-func (r *Reactor) AddPeer(peer p2p.Peer) {
-	r.mtx.RLock()
-	defer r.mtx.RUnlock()
-	if r.syncer != nil {
-		r.syncer.AddPeer(peer)
+// OnStop stops the reactor by signaling to all spawned go routines to exit and
+// blocking until they all exit. Finally, it will close the snapshot and chunk
+// p2p Channels respectively.
+func (r *Reactor) OnStop() {
+	r.snapshotChDone <- true
+	r.chunkChDone <- true
+	r.peerUpdatesDone <- true
```

Comment:
> I think we can probably simplify this by using the channel closure as a signal, and then simply doing `for e := range snapshotCh` in the goroutines. But we need to consider this in the context of `InCh` handling and closing in the shim, and I think we need to think a bit closer about this. Let's discuss.

This is originally how I had it implemented, however, I opted for explicit "done" channels because I came up with some corner-cases in terms of what is stopped first, the shim or real reactor, that could cause issues. I can't recall the exact corner-cases, it seems to escape me atm, but I'm open to discussing 👍 

So I would suggest we keep this explicit approach for now simply because it's safer and easier to reason about (at least for me.

> Also, if we do end up using done-channels, they should use channel closure as a signal since a closed channel always returns a value thus allowing multiple goroutines to use the same channel as a signal.

I would rather make this change if you're ok with it.

Reply:
I believe there are two problems with the current implementation:

1. The `Out` channel is closed by a different goroutine than the one that is sending on it. This means that the sender will panic e.g. if it is currently blocked on sending to the channel or if it has already taken a `select` branch and is about to send a message.

2. Not draining the `In` channel before exiting may cause the `In` sender to halt and/or leak goroutines.

This gets a bit tricky because these channels are bidirectional, and we can't really synchronize them to close at the same time. Also, we can't let the `In` consumer goroutine drain the channel, because doing so might cause it to send on `Out` which has already been closed, thus panicing.

I have a couple of ideas for how to deal with this, would love to hear your thoughts as well. Let's jump on a call later today?

Reply:
I do not see how (1) is possible. We do not close the `Out` channel until we've exited that goroutine (e.g. `processSnapshotCh`). But sure, we can close the `Out` channel upon exiting to be more explicit.

I also do not quite follow (2) -- the shim is the sender of the `In` channel and it does not spawn any goroutines.

But, yes, let's discuss 👍

Reply:
>I do not see how (1) is possible. We do not close the Out channel until we've exited that goroutine

Ah, you're right, because the send on that channel is synchronous of course. Sorry, I'm used to using `close()` for done-channels which is async.

>I also do not quite follow (2) -- the shim is the sender of the In channel and it does not spawn any goroutines.

The switch/MConnection spawns goroutines, and call `Receive` concurrently from these. The router will do something similar.

Reply:
Gave this pattern some thought over the weekend, here's a suggestion for some minor improvements:

* Use a single Go channel `closeCh` for signaling shutdown, which will be closed in `OnStop`.
* Check for `<-closeCh` in each goroutine, and then call `Channel.Close()` when the goroutine terminates (possibly with a `defer` so that it's robust to panics too).
* Close `closeCh` in `Stop` and wait for `Channel.Done()` on each channel.

This has a few advantages over the current pattern:

* The goroutine that sends on a channel also closes it, which follows best practices and is easier to reason about (less synchronization).
* This is resistant to panics, correctly signaling channel closure when the consumer dies regardless of termination conditions.
* Goroutines terminate in parallel, making shutdown faster if there is any significant cleanup to be done.
* `OnStop` blocks until the goroutines have actually returned, which may be significant if there is any cleanup to be done (otherwise the main process may e.g. exit before the shutdown completes, which would be bad if it involved e.g. saving data to disk).

Reply:
Pushed modifications.